### PR TITLE
fix: add code location and codeframe for rollup errors

### LINKED ISSUE #165: Add code location to rollup errors

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves #165

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR adds clickable code location and codeframe for rollup build errors.

---

CleanShot 2022-08-10 at 14 22 39

---

CleanShot 2022-08-10 at 21 38 46@2x

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/build.ts [hunk 1]
```diff
@@ -138,7 +138,8 @@ async function _build (nitro: Nitro, rollupConfig: RollupConfig) {
 
   nitro.logger.start('Building server...')
   const build = await rollup.rollup(rollupConfig).catch((error) => {
-    nitro.logger.error('Rollup error: ' + error.message)
+    const location = `${relative(process.cwd(), error.id)}:${error.loc.line}:${error.loc.column}`
+    nitro.logger.error(`Rollup error while parsing \`${location}\`` + '\n' + '\n' + error.frame)
```
Comment (pi0) [inline src/build.ts:142]: Is it only _parsing_ error possible?

Comment (danielroe) [inline src/build.ts:142]:
```suggestion
    nitro.logger.error(`Rollup error while processing \`${location}\`` + '\n' + '\n' + error.frame)
```

File: src/build.ts [hunk 2]
```diff
@@ -137,9 +137,19 @@ async function _build (nitro: Nitro, rollupConfig: RollupConfig) {
...
+      for (const error of ('errors' in _error ? _error.errors : [_error])) {
+        const id = error.id || _error.id
+        let path = isAbsolute(id) ? relative(process.cwd(), id) : id
+        const location = error.loc || error.location
+        if (location) {
+          path += `:${location.line}:${location.column}`
+        }
+        nitro.logger.error(`Rollup error while processing \`${path}\`` + '\n' + '\n' + (error.text || error.frame))
+      }
+    } catch {}
```
Comment (pi0) [inline src/build.ts:151]: Why another error should happen? If it is our internal code, it shouldn't be hidden like this IMO.

File: src/build.ts [hunk 3]
```diff
@@ -137,9 +137,19 @@ async function _build (nitro: Nitro, rollupConfig: RollupConfig) {
...
+  const build = await rollup.rollup(rollupConfig).catch((_error) => {
+    try {
+      for (const error of ('errors' in _error ? _error.errors : [_error])) {
+        const id = error.id || _error.id
+        let path = isAbsolute(id) ? relative(process.cwd(), id) : id
+        const location = error.loc || error.location
+        if (location) {
+          path += `:${location.line}:${location.column}`
+        }
+        nitro.logger.error(`Rollup error while processing \`${path}\`` + '\n' + '\n' + (error.text || error.frame))
```
Comment (pi0) [inline src/build.ts:149]: Maybe we can refactor this in a function utility `showRollupError` (could be even same file) and reuse for https://github.com/unjs/nitro/blob/0aa77a9ed8e125beab281e5b630990d38096e66f/src/build.ts#L207

**Review by pi0 — APPROVED** [review]
Thanks!

**Pull request merged**